File: binary_tree_set.scala
import scala.math.Ordering.Implicits.*

enum Tree[+A]:
  case Empty
  case Node(value: A, left: Tree[A], right: Tree[A])

def insert[A : Ordering](x: A, t: Tree[A]): Tree[A] =
  t match
    case Tree.Empty => Tree.Node(x, Tree.Empty, Tree.Empty)
    case Tree.Node(y, left, right) =>
      if x < y then Tree.Node(y, insert(x, left), right)
      else if x > y then Tree.Node(y, left, insert(x, right))
      else t

/**
 * Basic version of member.
 * This performs 2 comparisons at each node (x < y and x > y).
 */
def member[A : Ordering](x: A, t: Tree[A]): Boolean =
  t match
    case Tree.Empty => false
    case Tree.Node(y, left, right) =>
      if x < y then member(x, left)
      else if x > y then member(x, right)
      else true

/**
 * This version will take d+1 max comparisons. It is exercise 2.2 from 
 * Purely Functional Data Structures
 */
def member2[A : Ordering](x: A, t: Tree[A], candidate: Option[A]): Boolean =
  t match
    case Tree.Empty if candidate.isEmpty => false
    case Tree.Empty => candidate.get == x 
    case Tree.Node(y, left, right) =>
      if x >= y then member2(x, right, Some(y))
      else member2(x, left, candidate)

@main def treeTest(): Unit =
  val nums = List(5, 3, 8, 1, 4, 6, 9)
  val tree = nums.foldLeft(Tree.Empty: Tree[Int])((t, x) => insert(x, t))

  println(s"Searching for 4 (exists): ${member2(4, tree, None)}")
  println(s"Searching for 7 (missing): ${member2(7, tree, None)}")

File: zigtree/src/binary_tree_set.zig
const std = @import("std");

/// A purely functional binary tree set.
pub fn BinaryTreeSet(comptime T: type) type {
    return struct {
        pub const Node = struct {
            value: T,
            // We use optional constant pointers for immutability and the 'Empty' state
            left: ?*const Node = null,
            right: ?*const Node = null,
        };

        // Alias for the root type
        pub const Tree = ?*const Node;

        /// Inserts a value into the tree. Since it's purely functional, it returns
        /// a new tree, allocating new nodes where the path diverges.
        pub fn insert(allocator: std.mem.Allocator, x: T, t: Tree) !Tree {
            // If the tree is not empty, unwrap the pointer to 'node'
            if (t) |node| {
                if (x < node.value) {
                    const new_node = try allocator.create(Node);
                    new_node.* = .{
                        .value = node.value,
                        .left = try insert(allocator, x, node.left),
                        .right = node.right,
                    };
                    return new_node;
                } else if (x > node.value) {
                    const new_node = try allocator.create(Node);
                    new_node.* = .{
                        .value = node.value,
                        .left = node.left,
                        .right = try insert(allocator, x, node.right),
                    };
                    return new_node;
                } else {
                    return t; // Value already exists, return existing tree
                }
            } else {
                // Tree.Empty case
                const new_node = try allocator.create(Node);
                new_node.* = .{ .value = x };
                return new_node;
            }
        }

        /// Basic version of member (2 comparisons per node)
        pub fn member(x: T, t: Tree) bool {
            if (t) |node| {
                if (x < node.value) return member(x, node.left);
                if (x > node.value) return member(x, node.right);
                return true;
            }
            return false;
        }

        /// Optimized version of member taking d+1 max comparisons
        pub fn member2(x: T, t: Tree, candidate: ?T) bool {
            if (t) |node| {
                if (x >= node.value) {
                    return member2(x, node.right, node.value);
                } else {
                    return member2(x, node.left, candidate);
                }
            } else {
                if (candidate) |c| {
                    return c == x;
                }
                return false;
            }
        }

        pub fn member3(x: T, t: Tree) bool {
            var node = t;
            while (node) |curr| {
                const compare = std.math.order(x, curr.value); 
                switch (compare) {
                    .lt => node = curr.left,
                    .gt => node = curr.right,
                    .eq => return true,
                }
            }
            return false;
        }
    };
}
